### PR TITLE
libsql-client: move cloudflare workers to a separate feature

### PR DESCRIPTION
**Pull request by psarna**
This commit introduces a Connection trait, and a `workers::Connection` class implementing it.
The trait abstracts away common code, and the Cloudflare Workers backend is now guarded by an opt-in feature flag.

Creating connections can also be abstracted away in order to deduplicate the code even further, but it's left for a follow-up.

### CODE REVIEW

File: libsql-client/src/connection.rs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /// # Arguments
+    /// * `stmts` - SQL statements
+    /// ```
+    async fn transaction(
+        &self,
+        stmts: impl IntoIterator<Item = impl Into<Statement>>,
+    ) -> Result<Vec<QueryResult>> {
+        // TODO: Vec is not a good fit for popping the first element,
+        // let's return a templated collection instead and let the user
+        // decide where to store the result.
```
Comment (haaawk) [inline libsql-client/src/connection.rs:47]: Are we popping the first element? It seems that the first element is skipped and only the last element is popped, no?

Comment (haaawk) [reply libsql-client/src/connection.rs:47]: Is this comment still valid?

Comment (psarna) [reply libsql-client/src/connection.rs:47]: nah, we're no longer using vectors, so not valid, I'll remove it

File: libsql-client/src/workers.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        pass: impl Into<String>,
+    ) -> Self {
+        let username = username.into();
+        let pass = pass.into();
+        let url = url.into();
+        // Auto-update the URL to start with https:// if no protocol was specified
+        let url = if !url.contains("://") {
+            "https://".to_owned() + &url
+        } else {
+            url
```
Comment (haaawk) [inline libsql-client/src/workers.rs:34]: Would be cool to change `http://` prefix to `https://` automatically too.

File: libsql-client/Cargo.toml [hunk 3]
```diff
@@ -13,4 +13,9 @@ url = "2.3.1"
 base64 = "0.21.0"
 num-traits = "0.2.15"
 serde_json = "1.0.91"
-worker = "0.0.12"
+worker = { version = "0.0.12", optional = true }
+anyhow = "1.0.69"
+async-trait = "0.1.64"
+
+[features]
+workers_backend = ["worker"]
```
[the remark this answers is not in the document]
Comment (psarna) [reply libsql-client/Cargo.toml:21]: Yup, it's a user-facing feature. It's used in lib.rs:
```
#[cfg(feature = "workers_backend")]
pub mod workers;
```
and users can use it in workers environment to connect. It already powers https://country-counter.p-sarna.workers.dev/

File: libsql-client/src/workers.rs (hunk 2 above)
Comment (psarna) [reply libsql-client/src/workers.rs:34]: @honzasp explicitly asked not to do it in one of the past reviews and I agreed with the sentiment - turso.io indeed uses https://, but technically users could run their own local testing servers that run on http://

**Review by haaawk — APPROVED** [review]
LGTM. Do we have any README that needs updating to the new usage?

File: libsql-client/src/workers.rs (hunk 2 above)
Comment (haaawk) [reply libsql-client/src/workers.rs:34]: Fair enough. Maybe at least a warning would be good though.

**Comment by psarna** [comment]
yup, we do, and I forgot to update it :sweat_smile: will fix in a minute

File: libsql-client/Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (haaawk) [reply libsql-client/Cargo.toml:21]: right. I somehow missed that usage. Thanks!

**Comment by psarna** [comment]
rebased, removed the deprecated comment, updated the README

**Review by haaawk — APPROVED** [review]

**Comment by psarna** [comment]
bors merge

**Pull request merged**